magento/magento2
# Fix performance leak in salesrule collection

### Description
Refactored sql query that created a huge temporary table for each request, when a greater amount of salesrules and coupon codes exists in database. The sorting of this table took a lot of cpu time.
The statement now consists of two subselects that drill down the remaining lines as far as possible, so that the remaining temporary table is minimal and easily sorted.

### Fixed Issues
1. magento/magento2#19117: High Database Load for Sales Rule Validation

### Manual testing scenarios
1. Create an active salesrule with no coupon code (`A`)
1. Create an inactive salesrule with no coupon code (`B`)
1. Create an active salesrule with `coupon_type = 'autogenerated'` and generate 5000 codes (`C`)
1. Create an active salesrule with `coupon_type = 'specific'` (`D`)
1. Create an inactive salesrule with coupon code (`E`)
1. Add a product to the cart (qualified for all active salesrules)
1. Go to checkout/cart
    1. The subtotal for salesrules only lists label `A`
1. Add one of the coupon codes of `C` to the cart
    1. the coupon code can be applied
    1. The subtotal for salesrules lists label `A,C`
1. Remove the applied coupon code
    1. the coupon code can be removed
1. Add the coupon code `D` to the cart
    1. the coupon code can be applied
    1. The subtotal for salesrules lists label `A,D`
1. Add the coupon code `E` to the cart
    1. the coupon code can not be applied
1. Check out the order
    1. checkout succeeds
1. Configure rule `A` to `priority =  1`
1. Configure rule `D` to `priority =  2`
1. Add the product to a new quote 
    1. The subtotal for salesrules only lists label `A`
1. Apply Coupon for salesrule `D`
    1. The subtotal for salesrules lists label `A,D`
1. Configure rule `A` to `priority =  3`
1. Update cart items quantity
    1. The subtotal for salesrules lists label `D,A`

### Additional Information
We use this fix in a magento commerce environment

### Contribution checklist
